lib/state/settings/actions.js

```diff
@@ -52,6 +52,18 @@ export const setSortType = sortType => ({
   sortType,
 });
 
+export const setSearchSortType = searchSortType => ({
+  type: 'setSearchSortType',
+  searchSortType,
+});
+
+export const toggleSearchSortOrder = () => (dispatch, getState) => {
+  dispatch({
+    type: 'setSearchSortReversed',
+    searchSortReversed: !getState().settings.searchSortReversed,
+  });
```

Comment:
<s>here's a philosophical point of reflection: where is best to store the "toggle" state?

we can do it here in Redux middleware, which makes the UI components easier because they only need to call `toggle()`, but it complicates the Redux flows because the actions are no longer predictable.

we can do it in the UI and pass the next state, but that means we have to track it in the UI component and so complicates/duplicates it there.

one place I've come to accept is to use a hybrid of `mapStateToProps()` to get the existing state, `mapDispatchToProps()` to get the dispatch function, and occasionally, rarely, to use `mergeProps()` (also part of `connect()` but usually one we leave out) to keep the UI API clean and agnostic about the toggle state.

---

usually these complications are indications of a missing Eureka moment in a data structure somewhere. why is it hard for the UI to send the toggle state? (hint, boolean values should always raise our suspicion)

---

🤔 what if we stored `sortDirection` instead of `isReversed`?

```js
export const toggleSearchDirectionFrom = direction => ( {
  type: SEARCH_SET_DIRECTION,
  direction: direction === 'normal' ? 'reversed' : 'normal',
} )
```

### Manage it in the UI
not so bad because the state is simple enough and clear enough

```js
toggleSearchDirection = () => this.props.toggleSearchDirection( this.props.searchDirection )

<button onClick={ this.toggleSearchDirection } />

const mapStateToProps = ( { search: { direction } } ) => ( {
  searchDirection: direction,
} )

const mapDispatchToProps = dispatch => ( {
  toggleSearchDirection: direction => dispatch( toggleSearchDirectionFrom( direction ) ),
} )
```

### use `mergeProps` to remove the knowledge from the UI
more complicated `connect()` but more straightforward UI

```js
<button onClick={ this.props.toggleSearchDirection } />

const mapStateToProps = ( { search: { direction } } ) => ( {
  searchDirection: direction,
} )

const mapDispatchToProps = dispatch => ( {
  setSearchDirection: direction => dispatch( toggleSearchDirectionFrom( direction ) ),
} )

const mergeProps = ( stateProps, dispatchProps, ownProps ) => ( {
  ...ownProps,
  ...stateProps,
  ...dispatchProps,
  toggleSearchDirection: () => dispatchProps.toggleSearchDirection( ownProps.searchDirection ),
} )

connect( mapStateToProps, mapDispatchToProps, mergeProps )( Component );
```

Of course, that looks scary, especially if unfamiliar with it. I'm a fan of tracking this in the UI until it's clear we need more abstraction.  In any case, I still prefer keeping middleware behaviors outside of the action creators because of the pain I've had in trying to debug `redux thunk` stuff like this. In #1725 I recently proposed a new Redux middleware just for the purpose of executing the current search query (which is of course a bit more complicated because the search params are complicated). It's not that much more than Redux thunk actions but it preserves the predictability and audibility of our action sequence. [[adding middleware](https://github.com/Automattic/simplenote-electron/pull/1725/files#diff-e65c9a1430a959fb1c65a34167163c89R38), [defining middleware](https://github.com/Automattic/simplenote-electron/pull/1725/files#diff-d2dde7710fab6b285db73d1a1383dba5R13)]

### Middleware for predictability and introspection

```js
export const toggleSortOrder = () => ( {
  type: SORT_ORDER_TOGGLE,
} );

export const sortDirection = ( state = 'normal', { type, direction } ) =>
  SORT_ORDER_SET === 
```
</s>

**Eureka!**

See? it was way simpler than we were realizing. This is best handled in the reducer where all the data lives. 

```js
export const sortOrder = ( state = 'normal', { type } ) =>
  SORT_ORDER_TOGGLE === type
    ? ( state === 'normal' ? 'reverse' : 'normal' )
    : state;

export toggleSortOrder = () => ( {
  type: SORT_ORDER_TOGGLE,
} );

<button onClick={ this.props.toggleSortOrder } />

const mapStateToProps = ( { search: { sortOrder } } ) => ( {
  sortOrder,
} )

const mapDispatchToProps = ( {
  toggleSortOrder
} )
```